campaigns/workflows.md

```diff
@@ -0,0 +1,345 @@
+change.  The easiest way to check-point is to simply apply an operation
+such as `to_netcdf` or `to_zarr` on an Xarray Dataset.  However, unless
+other operations in the same workflow are aware of the check-point data,
+re-running the workflow will simply recompute the artifact and re-save
+it to disk.
+
+From a user's point of view, ideally, check-pointing should be automatic.
+That is, if a an object is computed in the notebook, it is cached on a
+persistent storage platform so that it can be retrieved from disk instead
+of recomputed.  This requires knowing the *history* of the artifact data
+and knowing if any element in the artifact's history has been modified
+(e.g., a modification to a function computing an intermediate product).
+If any element in the artifact's history has changed, then it must be
+re-computed "when accessed."
+
+To limit the amount of data stored on disk, it makes sense to only
+cache selected thing in the Notebook.  These items might correspond to
+the "public API" that a user designates for a notebook in the
+*Workflow Extension* model of [Search & Discovery](#search--discovery).
+
+#### Possible Projects
+
+- An automatic caching mechanism for the data objects in the "public API"
+  of a Notebook.  Could be similar to Xpersist, but could also be more
+  generic.  This is the same as the project suggestion in
+  [Section 1](#search--discovery).
+
+### Publication
+
+The publication phase of a workflow is when the final workflow is
+*published* for other users to use.  This is the same as the process
+suggested in the [Search & Discovery](#search--discovery) section,
+above.  This phase involves submitting the Notebook to a "repository"
+so that other users can "download" the Notebook for modification or
+import elements of the Notebook's "public API" into their own Notebook.
+
+Ideally, published Notebooks will be tested for regression as new
+versions of the software stack change, so that owners of the Notebook
+know that their Notebook will no-longer work with updated versions of
+the Notebook environment.
+
+Ideally, published Notebooks will be versioned and tied to a specific
+environment/kernel with pinned versions of the software used in the
+Notebook.  Ideally, the version of the Notebook and the version of the
+environment can be updated independently of each other.
+
+Ideally, published Notebooks will be published with metadata that
+contains (at a minimum) a description of what the Notebook does, a
+list of "public API" elements, the author's name and contact information
+(email), the version of the Notebook, the date and timestamp of the
+last change to the Notebook.  Ideally, this metadata will also contain
+information about who has "downloaded" the Notebook for modification,
+who has used which "public API" elements of the Notebook, how many
+times has the Notebook been "downloaded" and how many times has each
+"public API" element been accessed.
```

Comment:
Is this idea extensible to people publishing a workflow in order to meet journal publication requirements?  If so, it might be worth mentioning.

Reply:
Very good point!  I'll mention it.